### PR TITLE
Add support to copy an object that inherits from `Subject`

### LINKED ISSUE #694: Use of self.__class__ in __copy__ method of Subject
**🚀 Feature**
possibility to inherit torchio Subject

**Motivation**

In our current workflow we want to build on top of the torchio framework. We have subjects with 2 mandatory properties, a study name and a patient id. Therefore we have a subject class that looks like this:

```python
class Subject(tio.Subject):

    def __init__(self, study_name: str, patient_id: str, *args, **kwargs):
        kwargs['study_name'] = study_name
        kwargs['patient_id'] = patient_id
        super().__init__(*args, **kwargs)
```
**Pitch**

What I would like to see is that this method:

```python
    def __copy__(self):
        result_dict = {}
        for key, value in self.items():
            if isinstance(value, Image):
                value = copy.copy(value)
            else:
                value = copy.deepcopy(value)
            result_dict[key] = value
        new = Subject(result_dict)
        new.applied_transforms = self.applied_transforms[:]
        return new
```
is replaced by

```python
    def __copy__(self):
        result_dict = {}
        for key, value in self.items():
            if isinstance(value, Image):
                value = copy.copy(value)
            else:
                value = copy.deepcopy(value)
            result_dict[key] = value
        new = self.__class__(**result_dict)
        new.applied_transforms = self.applied_transforms[:]
        return new
```
to make custom subject classes that inherit torchio Subject compatible with the other components in the framework. The Image class has already a  __copy__ method that uses __class__ property.

**Alternatives**

The alternative is overwriting the __copy__ method in the custom classes that inherit torchio subject. However, this seems less future proof.

### PR DESCRIPTION
**Pull request by dmus**
Fixes #694.

**Description**
use of __class__ in __copy__ method of subject to make subclass Subject possible

**Checklist**

- [x] I have read the [`CONTRIBUTING`](https://github.com/fepegar/torchio/blob/main/CONTRIBUTING.rst) docs and have a developer setup (especially important are `pre-commit`and `pytest`)
- [x] Non-breaking change (would not break existing functionality)
- [ ] Breaking change (would cause existing functionality to change)
- [x] Tests added or modified to cover the changes
- [ ] Integration tests passed locally by running `pytest`
- [ ] In-line docstrings updated
- [ ] Documentation updated, tested running `make html` inside the `docs/` folder
- [x] This pull request is ready to be reviewed
- [ ] If the PR is ready and there are multiple commits, I have [squashed them and force-pushed](https://www.w3docs.com/snippets/git/how-to-combine-multiple-commits-into-one-with-3-steps.html#force-pushing-commits-7)

### CODE REVIEW

File: torchio/data/subject.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ def __copy__(self):
             else:
                 value = copy.deepcopy(value)
             result_dict[key] = value
-        new = Subject(result_dict)
+        new = self.__class__(**result_dict)
```
Comment (justusschock) [inline torchio/data/subject.py:74]: Why can't you simply use the dict as before here?

File: tests/data/test_subject.py [hunk 2]
```diff
@@ -106,3 +106,24 @@ def test_bad_arg(self):
...
+    def test_copy_subclass_no_kwargs(self):
+        class CustomSubject(tio.Subject):
+            def __init__(self):
+                tensor = torch.rand(1, 3, 3, 4)
+                img = tio.ScalarImage(tensor=tensor)
+                kwargs = {'img': img}
+                super().__init__(**kwargs)
+        subject = CustomSubject()
+        new_subject = copy.copy(subject)
+        assert isinstance(new_subject, CustomSubject)
```
Comment (justusschock) [inline tests/data/test_subject.py:129]:
```suggestion
    def test_copy_subclass(self):
        class CustomSubject(tio.Subject):
            pass
        assert isinstance(new_subject, CustomSubject)
```

IMO overything else (including argument forwarding) tests the correct implementation of the subclass which is up to the user

**Review by justusschock — APPROVED** [review]

**Comment by justusschock** [comment]
+1 on this.

@fepegar what do you think on my arguments on the tests above? For me they succeeded locally.

File: torchio/data/subject.py (hunk 1 above)
Comment (justusschock) [inline torchio/data/subject.py:74]:
```suggestion
        new = type(self)(**result_dict)
```
is also a bit cleaner :)

**Comment by fepegar** [comment]
I'm not sure this can be solved easily. @justusschock, if you have a branch with your proposed changes, we can test that one. I don't think we should remove the current tests, though.

**Comment by fepegar** [comment]
Maybe it would be easier to override the `__copy__` method, but I'm not sure.

**Comment by justusschock** [comment]
@fepegar I opened a similar PR with more minimal tests in #794

Overriding the `__copy__` could definitely work, but it would be a lot of copy-paste because you'd need to add all the other logic as well :)

**Comment by fepegar** [comment]
Closed in favor of #794.

**Pull request closed without merging**